platform/darwin/src/MGLMultiPoint.mm

```diff
@@ -24,105 +24,84 @@ - (instancetype)initWithCoordinates:(CLLocationCoordinate2D *)coords count:(NSUI
+    return _coordinates.at(0);
 }
 
 - (NSUInteger)pointCount
 {
-    return _count;
+    return _coordinates.size();
 }
 
 + (NS_SET_OF(NSString *) *)keyPathsForValuesAffectingPointCount
 {
     return [NSSet setWithObjects:@"coordinates", nil];
 }
 
+- (CLLocationCoordinate2D *)coordinates
+{
+    return _coordinates.data();
+}
+
 - (void)getCoordinates:(CLLocationCoordinate2D *)coords range:(NSRange)range
 {
-    if (range.location + range.length > _count)
+    if (range.location + range.length > [self pointCount])
     {
         [NSException raise:NSRangeException
                     format:@"Invalid coordinate range %@ extends beyond current coordinate count of %zu",
-                        NSStringFromRange(range), _count];
+                        NSStringFromRange(range), [self pointCount]];
     }
 
-    NSUInteger index = 0;
+    std::copy(_coordinates.begin() + range.location, _coordinates.begin() + NSMaxRange(range), coords);
+}
 
-    for (NSUInteger i = range.location; i < range.location + range.length; i++)
-    {
-        coords[index] = _coordinates[i];
-        index++;
-    }
+- (void)appendCoordinates:(CLLocationCoordinate2D *)coords count:(NSUInteger)count
+{
+    [self willChangeValueForKey:@"coordinates"];
+    _coordinates.insert(_coordinates.end(), *coords);
+    [self computeBounds];
+    [self didChangeValueForKey:@"coordinates"];
 }
 
 - (void)replaceCoordinatesInRange:(NSRange)range withCoordinates:(CLLocationCoordinate2D *)coords
 {
-    if ((coords >= _coordinates && coords < _coordinates + _count) ||
-        (coords + range.length >= _coordinates && coords + range.length < _coordinates + _count))
-    {
-        [NSException raise:NSRangeException format:@"Reuse of existing coordinates array %p not supported", coords];
-    }
-    else if (range.length == 0)
+    if (range.length == 0)
     {
-        [NSException raise:NSRangeException format:@"Empty coordinate range %@", NSStringFromRange(range)];
+        return;
     }
-    else if (range.location > _count)
+    
+    if (NSMaxRange(range) > _coordinates.size())
     {
         [NSException raise:NSRangeException
                     format:@"Invalid range %@ for existing coordinate count %zu",
-                        NSStringFromRange(range), _count];
+         NSStringFromRange(range), [self pointCount]];
     }
 
     [self willChangeValueForKey:@"coordinates"];
-    if (NSMaxRange(range) <= _count)
-    {
-        // replacing existing coordinate(s)
-        memcpy(_coordinates + range.location, coords, range.length * sizeof(CLLocationCoordinate2D));
-    }
-    else
+    std::copy(coords, coords + range.length, _coordinates.begin() + range.location);
+    [self computeBounds];
+    [self didChangeValueForKey:@"coordinates"];
+}
+
+- (void)computeBounds
+{
+    CLLocationCoordinate2D *coords = self.coordinates;
+    mbgl::LatLngBounds bounds = mbgl::LatLngBounds::empty();
+    for (NSUInteger i = 0; i < [self pointCount]; i++)
```

Comment:
This will call `-pointCount` on every iteration. Use C++ fast enumeration instead:

```cpp
for (auto coordinate : _coordinates) {
    …
}
```